game-core/src/main/java/games/strategy/engine/config/FilePropertyReader.java

```diff
@@ -39,27 +39,25 @@ public FilePropertyReader(final String path) {
   /**
    * Creates a property reader using the specified properties file as the source.
    *
-   * @param file The properties file.
-   *
    * @throws IllegalArgumentException If {@code file} does not exist.
```

Comment:
`{@code file}` -> `{@code propertiesFile}`

Reply:
Good catch. Looking at this a second time again, I'm tempted to remove it again. It's a bit odd to document unchecked exceptions, it is not the case that we would want someone to catch 'IllegalArgumentException'. The fact that a precondition state is asserted when we enter this method IMO does not need to be part of the public API documentation. Particularly since we're now seeing the maintenance cost, I'm pretty convinced it's better to remove these than to maintain them.

Reply:
Kind of disagree here.  Documenting preconditions on API (public/protected) methods is generally a Good Thing.  From _Effective Java_, Item 74:

> A well-documented list of the unchecked exceptions that a method can throw effectively describes the _preconditions_ for its successful execution. It is essential that every public method’s documentation describe its preconditions (Item 56), and documenting its unchecked exceptions is the best way to satisfy this requirement.

Reply:
I removed the @throws in: 6b9b3ef
Digging into this code, I noticed one constructor was test only, so I went further and removed that constructor and converted the tests to use the prod-code constructor, this eliminated some more of the javadoc

Reply:
I'm thinking about this from a few angles:

1. I don't think the spirt of item 74 is to spell out every NPE that is possible, or that InterruptedException is propagated, etc. By the same token we should document that a NPE is also possible if the arg were null.

There is this note though:

> It is fine for a class's document to say, "All methods in this class throw a `NullPointerException` if a null object reference is passed in any parameter".

We don't do that necessarily. For a `File` parameter, passing a file that does not exist and getting `FileNotFoundException` or `IllegalArgumentException` seems reasonable enough to have as blanket convention, just as much that if you pass a null arg you'll likely get a null arg exception.

2. The scope/intent of our javadocs, it's there for when you're reading a javadoc as part of an auto-complete, or when you're reading the code. We do not have the use-case to generate a javadoc html set for consumption independent of the code, I'm pretty confident we will not have that use-case considering the application is a thick client game. So documenting every time we use precondition checks for relatively straight forward assertions I think will have a negative impact:
- the overhead will discourage the practice
- it's maintenance burden

3. Getting more to this, I think the spirit is more for unchecked exceptions that are beyond precondition or NPE checks. Things that clients actually would want to catch, and in some cases *must* catch. In part that is symptom of control-flow-by-exception-handling, but for me the spirit of item 74 is more to document any exceptions used as part of control-flow, or exceptions which are more interesting than standard arg checks.


I'm considering what a good path forward or compromise would be. If we keep the documentation, it would imply that we should have such documentation everywhere. I don't think that should be the case:
- the exceptions thrown are error detection more than public API, there is no recovery mechanism as it's a coding bug if we hit this.
- documenting all precondition checks is just not going to be useful for us. I don't see it being helpful once compared to just looking at the first few lines of code where a precondition check lives. The comment on the precondition check is a good place to document this, then we get the benefit that if we actually see the failure we'll get information instead of regretting we did not copy/paste the javadoc into the error comment (that copy/paste is a DRY violation, seems like the place to document the precondition check is in the precondition check. That is in-line with javadoc which was originally invented to combine code and documentation to be in on place (keeping API docs together used to be a huge pain). I wonder what Bloch's take woudl be on Guava preconditions, and documenting those in the precondition check itself. The refactoring aspect is nice as we don't create a coupling between exception type and the documentation). 


------------------------


In this case we should be throwing `FileNotFound`, it's more specific to the actual error, the fact that this is implemented as a public library that is called internally so we get `IllegalArg` is really immaterial to the user, IllegalArgumentException would make no sense. This line of discussion is what makes me think that documenting such an error condition is over specified; I'd rather not have it, put some comments in the Precondition check if it makes sense, and keep things lightweight. On the other hand, if the unchecked exception were any more sophisticated or less conventional, I would be very much on board with documenting it. If we document it here, I think that places a very unhealthy and not useful standard on the rest of the code base.

Reply:
@ssoloff I'm having trouble adding these back in even to settle the dispute, example:

```

   * @throws IllegalArgumentException If {@code file} does not exist.
   */
  public FilePropertyReader(final File file) {
    checkNotNull(file);
    checkArgument(file.exists(), "Property file not found: " + file.getAbsolutePath());


```

In this case the javadoc is incomplete, it should also list the NPE check too. TBH this seems like busy-work for us. If we had more documentation of control-flow-by-exception handling, I'd be on board, but documenting effectively dead code put in place because at one time there was a preference to use guava precondition instead of the built-in assert, seems like we are doing it wrong.

Reply:
> By the same token we should document that a NPE is also possible if the arg were null.

I'm pretty sure the team has had this discussion several times, and the majority opinion has been types are assumed to be non-`null` by default (at least for new code), and we would only document the exceptional cases where `null` is permitted (preferably through the use of annotations).  #2885 is still open to determine if we should use `@Nonnull` explicitly or assume that it is implied by the absence of any nullity annotation (or another annotation at a higher scope that communicates such a default).  Therefore, I'd argue we already have an unwritten rule equivalent to Bloch's package example such that documenting NPEs is not required.

> the exceptions thrown are error detection more than public API, there is no recovery mechanism as it's a coding bug if we hit this.

That's pretty much the definition of a precondition violation: an unrecoverable programmer error that should not be caught.  I'm not sure a missing file falls under that definition.

> For a File parameter, passing a file that does not exist and getting `FileNotFoundException` or `IllegalArgumentException` seems reasonable enough to have as blanket convention,

For the case under discussion, I think the precondition check is probably not appropriate because it's basically a TOCTOU.  All that matters for correct operation of the class is that an exception is thrown if the file doesn't exist when we actually attempt to open it in `readProperty()`.

In that sense, a missing properties file is a recoverable error (and my preference is to communicate those types of errors using checked exceptione) because the caller may choose to fall back to a different file or a completely different mechanism to obtain the properties.